### PR TITLE
[AC-1124] Restrict admins from accessing items in the Collections tab

### PR DESCRIPTION
**Pull request by shane-melton**
## Type of change

```
- [ ] Bug fix
- [X] New feature development
- [ ] Tech debt (refactoring, code cleanup, dependency upgrades, etc)
- [ ] Build/deploy pipeline (DevOps)
- [ ] Other
```

## Objective

Restrict admins from accessing organization ciphers when the organization collection setting is disabled.

Required Server PR: https://github.com/bitwarden/server/pull/3676

### `libs/common/src/admin-console/models/domain/organization.ts`
- Add new `canEditAllCiphers()` helper to replace `canUseAdminCollections()` and add checks for `allowAdminAccessToAllCollectionItems` and providers.

### `vault/org-vault/vault.component.html`
- Add separate section for pre/post flexible collections V1
- The V1 section uses the `bit-no-items` component and the new collection access restricted component.

### `vault/org-vault/vault.component.ts`
- Deprecate the `showMissingCollectionPermissionMessage` as it is being replaced with `showCollectionAccessRestricted`
- Load `FlexibleCollectionsV1` feature flag on `ngOnInit()`

#### Cipher Changes
- Update `allCiphers$` to only load the ciphers available to the user depending on the feature flag and organization setting limiting admin access.
  - Ciphers are always loaded from the API now for the org vault, and do not use sync data to avoid mixing sync and API data in a single component and having to track which endpoint to use.
- Create an `allCipherMap$` to quickly look up cipher details by Id without needing to use `.find()` repeatedly. 
- `showCollectionAccessRestricted$` introduced and checks if the user has permission to view cipher items in the currently selected collection. 
- Update `ciphers$` observable to hide ciphers when being viewed in a collection the user does not have view permission for (normally the cipher isn't available, but if it is in multiple collections, one of which the user has access to, it should be hidden when the restricted collection is selected to avoid confusion)
- Update page route handling to check if the cipherId in the `route.queryParams` is editable before attempting to load and open the add-edit dialog. 
  - Must use the `allCipherMap$` as the cipher service sync data does not contain unassigned ciphers that may be available to admins. 

#### Collection Changes
- Create a list of `editableCollections$` that filters the list of all collections to only those that can be edited by the active user.
  - This is used when opening the add-edit and collections dialogs when admins are restricted. They will only be allowed to create ciphers in collections they have edit access to and can only move ciphers to collections they have edit access too.
  - Note: Currently, an admin will overwrite any other collection assignments that may exist if they do not have access to that collection. This will be addressed in a future ticket prior to V1 release.

### `libs/angular/src/vault/components/add-edit.component.ts`

Update save logic to use the correct cipher save endpoint based on the `orgAdmin` argument. We need to use the `organization.canEditAllCiphers()` helper to determine which endpoint to use.

### `libs/common/src/vault/services/cipher.service.ts`
- Add method to `getManyFromApiForOrganization` to only fetch ciphers the active user has access to, including unassigned ciphers for admins.

## Screenshots

Demo for an Admin user accessing the Org Collections tab with restricted access to some collection.

https://github.com/bitwarden/clients/assets/8764515/9af97ea8-29df-45be-b016-dae5d8e55340

## Before you submit

- Please add **unit tests** where it makes sense to do so (encouraged but not required)
- If this change requires a **documentation update** - notify the documentation team
- If this change has particular **deployment requirements** - notify the DevOps team
- Ensure that all UI additions follow [WCAG AA requirements](https://contributing.bitwarden.com/contributing/accessibility/)

### CODE REVIEW

File: libs/common/src/admin-console/models/domain/organization.ts [hunk 1]
```diff
@@ -196,6 +196,20 @@ export class Organization {
...
   }
 
+  canEditAllCiphers(flexibleCollectionsV1Enabled: boolean) {
+    // Before Flexible Collections, anyone with editAnyCollection permission could edit all ciphers
+    if (!flexibleCollectionsV1Enabled) {
+      return this.canEditAnyCollection;
+    }
+    // Post Flexible Collections V1, the allowAdminAccessToAllCollectionItems flag can restrict admins
+    // Providers are not affected by allowAdminAccessToAllCollectionItems flag
+    // note: canEditAnyCollection may change in the V1 to also ignore the allowAdminAccessToAllCollectionItems flag
```
Comment (vincentsalucci) [inline libs/common/src/admin-console/models/domain/organization.ts:206]: I apologize - but these comments confused me a bit. I just wanted to confirm that the **only** way to "Edit All Ciphers" (with the v1 flag **enabled**) is to the be a _provider_ or have admins allowed to access all collections **and** be either an _owner, admin, provider, or a custom user with the EditAnyCollection permission_. No other use cases allow a user access to all collections within an organization, right?

Comment (shane-melton) [reply libs/common/src/admin-console/models/domain/organization.ts:206]: I know we talked about this during a sync but wanted to put the answer here as well for others reading. Yes, those are the only expected use cases for a user to access all collections within an organization.

**Review by vincentsalucci — APPROVED** [review]
Reviewed AC owned file

**Review by JaredSnider-Bitwarden — DISMISSED** [review]
LGTM from auth perspective!

**Review by gbubemismith — DISMISSED** [review]

**Review by gbubemismith — DISMISSED** [review]

**Review by vincentsalucci — DISMISSED** [review]

**Review by JaredSnider-Bitwarden — APPROVED** [review]

**Review by vincentsalucci — APPROVED** [review]

**Review by gbubemismith — APPROVED** [review]

**Pull request merged**